### PR TITLE
WIP: allow changing context module in REPL

### PR DESCRIPTION
**Pull request by aviatesk**
### UPDATED

[After a discussion](https://github.com/JuliaLang/julia/pull/33102#issuecomment-525724114), I decided to add a functionality to change the REPL context module for execution, completion, and docs, in this PR

#### How works

The context module can be set via calling `REPL.set_context_module(mod)`, and then execution/completion/show docs/display are all done in the context of `mod`

#### TODO

- [x] add tests
    * change module in the context of execution
    * change module in the context of completion
    * change module in the context of help
- update docs ?: the method signature of `eval_user_input` has been changed and it appears in some docs like stack trace, and profiling. 

***

[The original description]

### Purpose

Allow REPL [tab] completions in a module context other than `Main`.

### Description

Currently REPL's [tab] completions work only in the context of the `Main` module, but there are cases when we want to get completions in a context of the other specific module:
E.g.: Juno can change REPL's module so that an entered code gets evaluated in a context of that module, and in this case we may want to get completions in the context of the module as well.

With this PR, each REPL provider could call `REPL.set_context_module(mod)` hooking on its module change, and then can get the corresponding completions.

If this idea is acceptable, I would really like to add tests for that, of course.

/cc @pfitzseb

### CODE REVIEW

**Comment by KristofferC** [comment]
This is already implemented by calling

```
completions(string, pos, context_module=Main)
```

https://github.com/JuliaLang/julia/blob/8093a7cec188f19782c6ad56bab6c7bcab060904/stdlib/REPL/src/REPLCompletions.jl#L573

See it in use in IJulia:

https://github.com/JuliaLang/IJulia.jl/blob/92b01a291519ecf68fbb58019c29b7747044584d/src/handlers.jl#L125

**Comment by aviatesk** [comment]
Thanks for the quick reply.
Yeah, I know, but that's not enough since tab completions are hooked at https://github.com/JuliaLang/julia/blob/master/stdlib/REPL/src/LineEdit.jl#L1826 and can't be changed outside of base as far as I understand.

**Comment by pfitzseb** [comment]
It would generally be useful to be able to set the REPL module (which Juno already allows you to do), both for execution and completions.
Only changing the context for completions on its own doesn't make too much sense, I think.

**Comment by aviatesk** [comment]
> It would generally be useful to be able to set the REPL module (which Juno already allows you to do), both for execution and completions.

It would make more sense to do so, and I'm willing to implement that if it sounds good.

AFAIU the current Juno's approach to change module context for REPL execution couldn't be applied to REPL completions (or could be done in a dirty way with a lot of insane copy and paste), thus I thought it would be better to be implemented here.

**Comment by aviatesk** [comment]
@KristofferC 
> But you mean that you want to be able to set that module globally for a REPL?

Changing module as for execution and completions (except ones via tab) has been done in lots of REPL provider but only tab completions haven't, thus I want to address that.

**Comment by pfitzseb** [comment]
AFAICT you only need to change [this line](https://github.com/JuliaLang/julia/blob/bb6c2686f891dd2442b53ed26fa4802f5a514d99/stdlib/REPL/src/REPL.jl#L86) plus some display logic below to use `context_module`.
That would still be a somewhat hacky solution though, and I think it would be better to pipe the current module [through the backend channel](https://github.com/JuliaLang/julia/blob/bb6c2686f891dd2442b53ed26fa4802f5a514d99/stdlib/REPL/src/REPL.jl#L705) instead, so that each request is associated with the correct module.

**Comment by aviatesk** [comment]
@pfitzseb 
> That would still be a somewhat hacky solution though, and I think it would be better to pipe the current module through the backend channel instead, so that each request is associated with the correct module.

Thanks for the pointers. Piping modules through `REPLBackend` sounds better.

@KristofferC 
So changing the module context for execution and completions (and maybe for helps) would be the preferred, after all ?
If so I'm going to change the PR title and get to the work.

**Comment by aviatesk** [comment]
Before proceeding to write tests, I would really like to hear your comments on those updates: https://github.com/JuliaLang/julia/pull/33102#issue-311819564
(for the general purpose of this PR and the current implementation approach or something :))

@KristofferC @pfitzseb

**Comment by aviatesk** [comment]
@pfitzseb @KristofferC @StefanKarpinski 
I finished up the work for now, and would like to hear your comments. The build failures seems to be irrelevant to the changes I made.

File: stdlib/REPL/src/docview.jl [hunk 1]
```diff
@@ -229,7 +229,8 @@ repl_search(s) = repl_search(stdout, s)
 
 function repl_corrections(io::IO, s)
     print(io, "Couldn't find ")
-    printstyled(io, s, '\n', color=:cyan)
+    pre = context_module == Main ? [] : [context_module, '.']
```
Comment (aviatesk) [inline stdlib/REPL/src/docview.jl]: Here I tried to mimic the behaviour introduced in #23806. It's not exactly same, but kind of consistent IMO.

File: stdlib/REPL/test/repl.jl [hunk 2]
```diff
@@ -1032,3 +1032,50 @@ fake_repl() do stdin_write, stdout_read, repl
     wait(repltask)
     @test istaskdone(repltask)
 end
+
+# context_module changing #33102
```
Comment (aviatesk) [inline stdlib/REPL/test/repl.jl]: I'm feeling like I want to add more tests here, and would really appreciate if you give me your ideas to make this test more effective.

**Review by pfitzseb — COMMENTED** [review]
IMHO this looks good in general, but a couple of things to think about:

- The whole `context_module` global is a bit messy as is -- I think it would be a bit cleaner (and much more complex) to *only*  use the global in the frontend and put the current module into `REPLDisplay` and `REPLCompletionProvider`, so that those don't rely on any globals. Haven't thought this through in detail, so it might a) not be much better and b) infeasible ;)
- It would be cool to have a minimal REPL mode that allows changing the current module without weird `Main.Base.set_context_module` invocations (which would be necessary in a `baremodule`). That mode could also display the current module.
- Should we change the `julia>` prompt to show the current module? Or change its color when you're not in `Main`? I feel like people are going to be confused without *some* indication that you're not in the default module.

**Comment by KristofferC** [comment]
Note that we have this global object to hold options:

https://github.com/JuliaLang/julia/blob/a9d4eacabae230d4ba5500078d50afc1eade2b57/stdlib/REPL/src/REPL.jl#L263-L311

Perhaps the `context_module` would fit in there? Not sure.

**Comment by pfitzseb** [comment]
So the reason why it's dangerous to have `context_module` as a global is that it can change at any time (e.g. when you switch the module in Juno), even between the backend evaluating a request and the result of that being displayed. To get this fully correct we can *only* access the global when initiating the request in the frontend, and then everything else (eval, completions, display) needs to use that, regardless of what value the global has now.

File: stdlib/REPL/src/REPL.jl [hunk 3]
```diff
@@ -59,6 +59,11 @@ answer_color(::AbstractREPL) = ""
 
 const JULIA_PROMPT = "julia> "
 
+context_module = Main
```
[the remark this answers is not in the document]
Comment (aviatesk) [reply stdlib/REPL/src/REPL.jl]: Definitely, I just pushed [the commit](https://github.com/JuliaLang/julia/pull/33102/commits/0cccfcf1ead2630c56a231e25268e71146b93cac) to use `Ref` for `context_module` :)

**Comment by aviatesk** [comment]
I think this is superseded by https://github.com/JuliaLang/julia/pull/33872.

**Comment by rfourquet** [comment]
Sorry for having started #33872 without checking (or finding) that there was already another attempt...

**Pull request closed without merging**